pkg/ccl/changefeedccl/changefeed_dist.go

```diff
@@ -261,16 +261,33 @@ func startDistChangefeed(
 	dsp := execCtx.DistSQLPlanner()
 
 	//lint:ignore SA1019 deprecated usage
-	var checkpoint *jobspb.ChangefeedProgress_Checkpoint
+	var legacyCheckpoint *jobspb.ChangefeedProgress_Checkpoint
 	if progress := localState.progress.GetChangefeed(); progress != nil && progress.Checkpoint != nil {
-		checkpoint = progress.Checkpoint
+		legacyCheckpoint = progress.Checkpoint
 	}
 	var spanLevelCheckpoint *jobspb.TimestampSpansMap
 	if progress := localState.progress.GetChangefeed(); progress != nil && progress.SpanLevelCheckpoint != nil {
 		spanLevelCheckpoint = progress.SpanLevelCheckpoint
 	}
+	if legacyCheckpoint != nil && spanLevelCheckpoint != nil {
+		if legacyCheckpoint.Timestamp.After(spanLevelCheckpoint.MinTimestamp()) {
+			// We should never be writing the legacy checkpoint again once we
+			// start writing the new checkpoint format. If we do, that signals
+			// a missing or incorrect version gate check somewhere.
+			return errors.AssertionFailedf("both legacy and current checkpoint set on " +
+				"changefeed job progress and legacy checkpoint has later timestamp")
+		}
+		// This should be an assertion failure but unfortunately due to a bug
+		// that was included in earlier versions of 25.2 (#148620), we may fail
+		// to clear the legacy checkpoint when we start writing the new one.
+		// We instead discard the legacy checkpoint here and it will eventually be
+		// cleared once the cluster is running a newer patch release with the fix.
+		log.Warningf(ctx, "both legacy and current checkpoint set on changefeed job progress; "+
+			"discarding legacy checkpoint")
+		legacyCheckpoint = nil
```

Comment:
reading makePlan i'm a little concerned that this could result in a nil ptr panic. do you think we have test coverage for this?

Reply:
It should be fine given the legacy checkpoint will be nil most of the time (i.e. when the bug doesn't happen)